### DISCUSSION TITLE
コードブロック内のプログラムの行数制限（質問）と　行番号表示（要望）

### DISCUSSION THREAD

**Discussion by tsubolabo** [increments/qiita-discussions#935] [要望・アイデア・不具合]
自分の過去の記事では、最大で3000行近いプログラムを添付したことがあります。しかし、普通は、1コードブロック当たり、300行前後に収まるように努力しています。それでも、他の記事と比較すると多めに感じられるので、投稿のたびに気になっています。
　現在作成中の記事では、更に多くの行数が必要になりそうなのですが、Qiitaで許容できる妥当な行数はどの程度なのでしょうか？

　現在、コードブロックでは行番号表示はできないようですが、オプションを指定することで表示できるようにはならないでしょうか？　「コードブロックの〇行目から△行目に示すように・・・」という説明もできるので、書き手にも読み手にも便利になると思います。最近は、Qiitaの投稿用のエディタでも、行番号が表示されるようになったので便利です。

**Comment by acronhub** [comment]
@tsubolabo 
> 現在作成中の記事では、更に多くの行数が必要になりそうなのですが、Qiitaで許容できる妥当な行数はどの程度なのでしょうか？

許容できる妥当な行数ですが、読み手によって妥当のラインはバラバラだと思うので断定することは難しいです。
恐れ入りますが、コードの内容や目的に応じて、適切な長さを判断いただければと思います :bow:

> 現在、コードブロックでは行番号表示はできないようですが、オプションを指定することで表示できるようにはならないでしょうか？

リクエストありがとうございます :pray:
現状、オプションを指定などでコードブロックに行番号を表示することはできないです
貴重なご意見として今後の参考にさせていただきます :pray:

代替案ですが、
重要な行に対してコメントで番号を振って、後で説明を書いたりするのはいかがでしょうか？

**Comment by TETSURO1999** [comment]
@tsubolabo
@acronhub

自分の場合，一画面に収まる程度であればコードをそのまま掲載しますが，コードが長くなり過ぎると記事の全貌が見えにくくなるので「折りたたみ」機能を使って隠すようにしています。

```markdown
<details><summary>コードは下記参照</summary>
ここに長文のコードを記載する
</details>
```

また行番号については自分も欲しいと感じる場合があります。C言語なら下記のように行頭に行番号をコメントで入れるとかでしょうか？

```c
/*0001*/#include <stdio.h>
/*0002*/int main() {
/*0003*/    printf("Hello World.\n");
/*0004*/    return 0;
/*0005*/}
```

ただしC言語は複数行に渡るコメントが可能なので，上記のように単純に行頭にコメントを入れるだけだと下記のケースでは文法エラーになってしまいます。

```c
/*
複数行に渡るコメント
*/
```

代替案として，特定の行を目立たそうとしたら下記のように `diff` 形式を利用する方法が考えられます。Qiita の markdown は `diff` 形式が充実しており，`diff_c`  形式などを指定できます。

```diff
 #include <stdio.h>
 int main() {
+    printf("Hello World.\n");
     return 0;
 }
```

**Comment by tsubolabo** [comment]
@acronhub 

ご返信ありがとうございます。

>コードの内容や目的に応じて、適切な長さを判断いただければと思います

　具体的な行数制限はないようなので、ひとまず安心しました。あまり冗長にならないように気を付けながら、より多くの情報を盛り込めるように対応させていただくことにします。

>オプションを指定などでコードブロックに行番号を表示することはできないです　貴重なご意見として今後の参考にさせていただきます

　宜しくお願いいたします。代替案までお示しいただきありがとうございます。提案が実現するまでの参考にさせていただきます。

@TETSURO1999 

色々な対策をお教えいただきありがとうございます。

>「折りたたみ」機能を使って隠すようにしています。

　「長すぎるのでは？」という後ろめたさも軽減できるので、それが良さそうですね。

>C言語なら下記のように行頭に行番号をコメントで入れるとかでしょうか？
>特定の行を目立たそうとしたら下記のように diff 形式を利用する方法が考えられます。

　ありがとうございます。参考にさせていただきます。ただ、折角のご提案にお言葉を返すようで恐縮ですが、個人的には次のような理由もあり、直ぐに利用するのは難しそうです。申し訳ありません。

①　私が良く使う言語では、一行の途中の一部だけをコメント化することはできないようです（コメント記号があると、行末までの全てがコメント扱いになる）。

②　なぜか、コメント行は淡い色で表示されるので、目印としては目立たちません。

③　Qiitaへの掲載のためだけに、本来はそのままでも動作するプログラムを書き直さなければならず、何らかの間違いが紛れ込みそうです。

④　コメント以外の文字を追加すると、コピーしたプログラムがそのままでは動作しなくなります。

　やはり、行番号の追加に期待しております。なお、上記④のことを考えると、Qiitaさんには、コードブロックをコピーしても、行番号は除外されるように配慮いただけると助かります。

**Reply by TETSURO1999** [reply]
@acronhub 

Qiita ではコードブロックのシンタクスハイライトを Ruby の Rougify という技術を使っているはずです。できれば Rougify 本体を弄らないで，その周辺で行番号を表示できるようしたい。とすれば，下記のように CSS を使って行番号を表示するというのはどうでしょうか？行番号の表示・非表示の切り替えもできそうです。

[CSSでコードブロックの行番号を表示する - Qiita](https://qiita.com/Ria0130/items/b49b13e4ff935993c813)

**Reply by acronhub** [reply]
@TETSURO1999
対策の共有ありがとうございます！
運営としても参考になります :bow: 

@tsubolabo
@TETSURO1999
直近の開発優先度から時期はお約束できませんが、社内で検討いたします。
ご提案ありがとうございます！😃
